### PR TITLE
Feat: add minting app page

### LINKED ISSUE #25: Minting website generator
## What we want

Currently the repo permits deploying an SBT contract like so, but in order for this site to be a Real no-code, users should also be able to create their own minting app. For the moment we'll keep it simple, the idea is to keep the same front-end that the first SBT app ([hackerhouse.starknet.id](https://hackerhouse.starknet.id)) but for any SBT contract. 

## How we see it

You'll need to create a page that can read the data of an SBT contract and show the image, and verify the password automatically. So you'll have the in query params the address of the SBT contract and you'll be able to get all the info needed to build your page. It should show if the mint is closed and the image of the SBT you're minting, the main difference that we'll have with hackerhouse.starknet.id is that we won't need any password here. 

The private key could be generated on the front end as we did in hackerhouse.starknet.id, at some point we'll use the wallet connected to do so.

I think this PR could take around 2-3 days depending on your level and your knowledge about the SBT contract.

## More context and contact

[Here is the repo](https://github.com/starknet-id/hackerhouse.starknet.id) of hackerhouse.starknet.id
[Here is the repo](https://github.com/starknet-id/identity_tokens) of the SBT contract it'll soon change but it's not a problem.

### PR DESCRIPTION
**Pull request by irisdv**
Closes #25 

Url should be formatted : 
- `sbt.starknet.id/{contractAddress}` for mainnet 
- `goerli.sbt.starknet.id/{contractAddress} `for testnet

### CODE REVIEW

File: src/pages/[SbtAddress].tsx [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+  calldata: (string | number)[];
+  contractAddress: string;
+  entrypoint: string;
+};
+
+type TransactionDatas = {
+  status: string;
+};
+
+const SbtGenerator: NextPage = () => {
```
Comment (fricoben) [inline src/pages/[SbtAddress].tsx]: It's a bit weird that the page name is `SbtAddress` and the component itself is named `SbtGenerator` don't you think ? 

to me it should be something like `sbtMintingPage` on both

File: src/pages/[SbtAddress].tsx [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+  useEffect(() => {
+    if (!loadingMessage) setElement(null);
+    setElement(<LoadingScreen message={loadingMessage} />);
+  }, [loadingMessage]);
+
+  useEffect(() => {
+    if (!loading) setElement(null);
+    if (address && loading && !error) setLoadingMessage("Fetching SBT data");
+  }, [loading, address]);
```
Comment (fricoben) [inline src/pages/[SbtAddress].tsx]: I'll test to see how it returns in the front but I feel like you don't need these 2 `useEffect`, only one with a skeleton for example.

Comment (fricoben) [reply src/pages/[SbtAddress].tsx]: I can't test cause it does not build as I told you on Telegram

File: src/pages/[SbtAddress].tsx [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    }
+  }, [transactionData, transactionLoading, transactionError]);
+
+  useEffect(() => {
+    if (!sbtData && data) {
+      let dataUri = "";
+      data[0].forEach((c: number) => {
+        dataUri += String.fromCharCode(c);
+      });
+      const metadataUri = dataUri.replace("ipfs://", "https://ipfs.io/ipfs/");
```
Comment (fricoben) [inline src/pages/[SbtAddress].tsx]: is this the best way to access the data ? 

Cause using `ipfs.io/ipfs` is very slow in general. If you don't know any other it's not worth searching atm :).

File: src/pages/[SbtAddress].tsx [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+                </>
+              )
+            ) : (
+              <>
+                <h1 className={styles.title}>
+                  {sbtData && !loading && !error
+                    ? sbtData?.name
+                    : loading && !error
+                    ? "Fetching SBT data..."
+                    : "SBT not found"}
```
Comment (fricoben) [inline src/pages/[SbtAddress].tsx]: A bit complicated to read maybe you could create a const for that.

File: src/pages/[SbtAddress].tsx [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+                </h1>
+                <p className={styles.text}>
+                  {sbtData
+                    ? sbtData?.description
+                    : !loading && error
+                    ? `We could not find the SBT at ${contractAddress} address, please try another one.`
```
Comment (fricoben) [inline src/pages/[SbtAddress].tsx]: Same here a bit complicated to read

File: src/pages/[SbtAddress].tsx [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+                      </div>
+                    </>
+                  ) : null}
+                </div>
+              </>
+            )}
+          </div>
+        </div>
+        <Connect />
+        {element}
```
Comment (fricoben) [inline src/pages/[SbtAddress].tsx]: I guess this is for the loading part, It's better to set a boolean and use `isLoading ? <Element /> : null` don't you think ? 

If there is a specific reason I would love to have an explanation :)

File: src/styles/generator.module.css [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+  flex-direction: column;
+  justify-content: center;
+  align-items: center;
+  background-color: var(--background);
+  max-width: 100vw;
+  height: 100vh;
+  /* background-image: url("/telAviv.webp");
+    background-size: cover;
+    background-repeat: no-repeat;
+    position: relative; */
```
Comment (fricoben) [inline src/styles/generator.module.css]: Delete this comment if it's not needed or useful, if it is add a explaination :)

File: src/styles/home.module.css [hunk 8]
```diff
@@ -33,15 +33,15 @@
...
+/* .formSideImage {
   width: 300px;
   height: 100%;
   border-radius: 20px;
   background-color: var(--brown);
   background-image: url("/visuals/sbtIllustration.webp");
   background-size: cover;
   background-position: center;
-}
+} */
```
Comment (fricoben) [inline src/styles/home.module.css]: Same here 🤟

**Review by fricoben — CHANGES_REQUESTED** [review]
Some comments 👀

File: src/pages/[SbtAddress].tsx (hunk 3 above)
Comment (irisdv) [reply src/pages/[SbtAddress].tsx]: Usually I use Pinata to fetch but recently I got errors when fetching metadata through it. Can retry Pinata and see how it goes

File: src/pages/[SbtAddress].tsx (hunk 6 above)
Comment (irisdv) [reply src/pages/[SbtAddress].tsx]: It's done that way on the page to deploy a SBT contract, that's why I did the same, but I can change it

File: src/styles/globals.css [hunk 9]
```diff
@@ -12,6 +12,8 @@
   --light-soft-brown: #f2ddbd;
   --beige: #fff9f0;
   --nimiq-ease: cubic-bezier(0.25, 0, 0, 1);
+  --title: #080161;
+  --action: #ff4f08;
```
Comment (fricoben) [inline src/styles/globals.css]: A ne pas utiliser car c'est les couleurs de starknet mais pas de starknet ID ;).

File: src/pages/[SbtAddress].tsx (hunk 1 above)
Comment (irisdv) [reply src/pages/[SbtAddress].tsx]: `SbtMintingPage` is much better, I replaced it
You mean also replacing `[SbtAddress]` by `SbtMintingPage` ? the page has a dynamic route which is based on the SBT contractAddress that's why I name it like this

Comment (fricoben) [reply src/pages/[SbtAddress].tsx]: Yeah I'm not sure what's the good practice here but the naming `SbtMintingPage` on both is better to me

File: src/pages/[SbtAddress].tsx (hunk 6 above)
Comment (fricoben) [reply src/pages/[SbtAddress].tsx]: Oh ok, @Marchand-Nicolas stuff haha. 

Can you change it on both ?

File: src/components/UI/textField.tsx [hunk 10]
```diff
@@ -31,7 +30,7 @@ const TextField: FunctionComponent<TextFieldProps> = ({
       helperText={helperText}
       defaultValue={defaultValue}
       variant="outlined"
-      onChange={onChange || (() => {})}
```
Comment (fricoben) [inline src/components/UI/textField.tsx:34]: ptdr ça n'avait tellement pas de sens 😭

**Review by fricoben — APPROVED** [review]

**Pull request merged**